src/System.Private.CoreLib/shared/System/MemoryExtensions.cs

```diff
@@ -22,6 +23,499 @@ namespace System
     /// </summary>
     public static partial class MemoryExtensions
     {
+        /// <summary>
+        /// Removes all leading and trailing occurrences of a specified element.
+        /// </summary>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        public static Memory<T> Trim<T>(this Memory<T> memory, T trimElement)
+            where T : IEquatable<T>
+        {
+            Span<T> span = memory.Span;
+            int start = ClampStart(span, trimElement);
+            int length = ClampEnd(span, start, trimElement);
+            return memory.Slice(start, length);
```

Comment:
There's a fair amount of duplication between the Memory and ReadOnlyMemory overloads.  Since this is all part of the same implementation, I'm wondering if we should just delegate from one to the other, and rely on MemoryMarshal.AsMemory to smooth things over, e.g.
```C#
public static Memory<T> Trim<T>(this Memory<T> memory, T trimElement) where T : IEquatable<T> =>
    MemoryMarshal.AsMemory(Trim((ReadOnlyMemory<T>)memory, trimElement));
```

Reply:
(Actually a question, not a request. :))

Reply:
I didn't know you could do that translation.
Do you mind if we merge this, then I will bench & PR the "fix" separately if it makes sense ?

Reply:
Microbench done. It looks like marshalling has an overhead:

```
|             Method |      Mean |     Error |    StdDev | Ratio | RatioSD | Rank |
|------------------- |----------:|----------:|----------:|------:|--------:|-----:|
|            TrimRoM | 0.0008 ns | 0.0000 ns | 0.0000 ns |  1.00 |    0.00 |    1 |
|         TrimMemory | 0.0007 ns | 0.0000 ns | 0.0000 ns |  0.99 |    0.03 |    1 |
| TrimMemoryMarshall | 0.0011 ns | 0.0000 ns | 0.0000 ns |  1.45 |    0.04 |    2 |
```

Reply:
`MemoryMarshal.AsMemory` and `Memory<T> implicit operator ReadOnlyMemory<T>` might need an aggressive inline; probably look like chunky methods to the Jit rather than something to evaporate.

Reply:
> might need an aggressive inline

You're probably right; I'll take a look.

Reply:
> might need an aggressive inline

I doubt it. I think that the problem is that the unsafe casts in these methods force the JIT to spill everything to stack and make copies.

Reply:
@jkotas is correct; it is getting inlined.